framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,874 @@
+            QualifierKind qualifierKind, Map<QualifierKind, Set<QualifierKind>> directSuperMap) {
+        Queue<QualifierKind> queue = new ArrayDeque<>(directSuperMap.get(qualifierKind));
+        Set<QualifierKind> allSupers = new TreeSet<>(directSuperMap.get(qualifierKind));
+        while (!queue.isEmpty()) {
+            QualifierKind superQual = queue.remove();
+            if (superQual == qualifierKind) {
+                throw new UserError("Cycle in hierarchy: %s", qualifierKind);
+            }
+            queue.addAll(directSuperMap.get(superQual));
+            allSupers.addAll(directSuperMap.get(superQual));
+        }
+        return allSupers;
+    }
+
+    /**
+     * Creates a mapping from {@link QualifierKindPair} to the least upper bound of both
+     * QualifierKinds.
+     *
+     * @return a mapping from {@link QualifierKindPair} to their lub.
+     */
+    protected Map<QualifierKindPair, QualifierKind> createLubsMap() {
+        Map<QualifierKindPair, QualifierKind> lubs = new TreeMap<>();
+        for (QualifierKind qual1 : nameToQualifierKind.values()) {
+            for (QualifierKind qual2 : nameToQualifierKind.values()) {
+                if (qual1.top != qual2.top) {
+                    continue;
+                }
+                QualifierKind lub = findLub(qual1, qual2);
+                QualifierKindPair pair = new QualifierKindPair(qual1, qual2);
+                QualifierKind otherLub = lubs.get(pair);
+                if (otherLub != null) {
+                    if (otherLub != lub) {
+                        throw new BugInCF(
+                                "Multiple lubs for qualifiers %s and %s. Found lubs %s and %s",
+                                qual1, qual2, lub, otherLub);
+                    }
+                } else {
+                    lubs.put(pair, lub);
+                }
+            }
+        }
+        return lubs;
+    }
+
+    /**
+     * Returns the least upper bound of {@code qual1} and {@code qual2}.
+     *
+     * @param qual1 a qualifier kind
+     * @param qual2 a qualifier kind
+     * @return the least upper bound of {@code qual1} and {@code qual2}
+     */
+    private QualifierKind findLub(QualifierKind qual1, QualifierKind qual2) {
+        if (qual1 == qual2) {
+            return qual1;
+        } else if (qual1.isSubtype(qual2)) {
+            return qual2;
+        } else if (qual2.isSubtype(qual1)) {
+            return qual1;
+        }
+        Set<QualifierKind> allSuperTypes = new TreeSet<>(qual1.superTypes);
+        allSuperTypes.retainAll(qual2.superTypes);
+        Set<QualifierKind> lubs = findLowestQualifiers(allSuperTypes);
+        if (lubs.size() != 1) {
+            throw new BugInCF(
+                    "Not exactly 1 lub for %s and %s. Found lubs: [%s].",
+                    qual1, qual2, SystemUtil.join(", ", lubs));
+        }
+        QualifierKind lub = lubs.iterator().next();
+        if (lub.isPoly && !qual1.isPoly && !qual2.isPoly) {
```

Comment:
Why isn't this an error?  (All comments about lub apply to corresponding glb methods too.)